resources/assets/ts/chat/main-view.tsx

```diff
@@ -0,0 +1,108 @@
+/**
+ *    Copyright 2015-2018 ppy Pty. Ltd.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { ChatChannelSwitchAction } from 'actions/chat-actions';
+import Dispatcher from 'dispatcher';
+import HeaderV3 from 'header-v3';
+import { observer, Provider } from 'mobx-react';
+import Channel from 'models/chat/channel';
+import User from 'models/user';
+import * as React from 'react';
+import RootDataStore from 'stores/root-data-store';
+import { ChannelJSON, UserJSON } from './chat-api-responses';
+import ChatWorker from './chat-worker';
+import ConversationList from './conversation-list';
+import ConversationView from './conversation-view';
+import InputBox from './input-box';
+
+interface PropsInterface {
+  dataStore: RootDataStore;
+  dispatcher: Dispatcher;
+  worker: ChatWorker;
+  presence: ChannelJSON[];
+}
+
+interface SendToJSON {
+  target: UserJSON;
+  can_message: boolean;
+}
+
+@observer
+export default class MainView extends React.Component<PropsInterface, any> {
+  constructor(props: any) {
+    super(props);
+
+    if (!_.isEmpty(props.presence)) {
+      this.props.dataStore.channelStore.updatePresence(props.presence);
+    }
+  }
+
+  componentDidMount() {
```

Comment:
most (all?) of this seem like they should be in the constructor or set up before `MainView` instead of after it's rendered once